### ISSUE TITLE
[docs] Document leaf node distinction in FAQ or Tutorials

### ISSUE THREAD

**Issue by GZHermit** [pytorch/pytorch#10065]
everyone, I find a problem about the change of the tensor's type.
This is my code:
```
w = torch.tensor([1, 2, 3], requires_grad=True)
w1 = w.type(torch.float)
print(w1, w1.type(), id(w1), w1.requires_grad)
o = torch.mean(w1)
o.backward()
print(w1.grad)
```
Why the w1.grad is None? And if you change the tensor's type by using the type_as, it will appear the same problem.

I need help...thank you

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @jlin27 @mruberry

**Comment by matthew-z** [comment]
I think it is a intended feature. dtype casting is not a differentiable op.

**Comment by zou3519** [comment]
@GZHermit ,

> w = torch.tensor([1, 2, 3], requires_grad=True)

is a leaf tensor, while the next line,

> w1 = w.type(torch.float)

creates a non-leaf tensor `w1`. Gradients only accumulate in leaf tensors (i.e., `w` in this case).

To fix the problem, you can do the following:

```
w1 = torch.tensor([1, 2, 3], requires_grad=True, dtype=torch.float).

I'll leave this issue open so that we will add a note to the FAQ and/or tutorials about this point.

**Comment by albanD** [comment]
Note that if you do this now, you will get the following warning when accessing the `.grad` field of w1:
```
UserWarning: The .grad attribute of a Tensor that is not a leaf Tensor is being accessed. Its .grad attribute won't be populated during autograd.backward(). If you indeed want the gradient for a non-leaf Tensor, use .retain_grad() on the non-leaf Tensor. If you access the non-leaf Tensor by mistake, make sure you access the leaf Tensor instead. See github.com/pytorch/pytorch/pull/30531 for more informations.
```

@zou3519 is that enough?

**Comment by zou3519** [comment]
@albanD that looks pretty good and explains the problem and solutions! Closing this issue
